// @vdian/date/lib/constants/index.js
// 根据符号匹配对应年月日时分秒的数字,通过构造函数实例化正则表达式
export var TOKEN_REGEX_MAP = {
    yy: '(\\d{2})',
    yyyy: '(\\d{4})',
    M: '((?:1[0-2])|[1-9])',
    MM: '((?:0[1-9])|(?:1[0-2]))',
    d: '((?:[1-2]\\d)|(?:3[0-1])|[1-9])',
    dd: '((?:0[1-9])|(?:[1-2]\\d)|(?:3[0-1]))',
    H: '((?:1\\d)|(?:2[0-3])|\\d)',
    HH: '((?:[0-1]\\d)|(?:2[0-3]))',
    m: '((?:[1-5]\\d)|\\d)',
    mm: '([0-5]\\d)',
    s: '((?:[1-5]\\d)|\\d)',
    ss: '([0-5]\\d)'
};
// 匹配一个正则表达式中的元字符
export var REGEX_SPECIAL_CHARACTER = /[-\/\\^$*+?.()|[\]{}]/g;
// 匹配一个用来解析时间的格式化符号
export var PARSE_TOKEN = /yyyy|yy|MM|M|dd|d|HH|H|mm|m|ss|s/g;
// 匹配一个用来格式化时间的格式化符号
export var FORMAT_TOKEN = /yyyy|yy|MM|M|dd|d|HH|H|mm|m|ss|s|E/g;
// 格式化日期时 默认的格式化参数
export var DEFAULT_FORMAT = 'yyyy-MM-dd';
// 解析时间时，默认的年
export var DEFAULT_YEAR = 0;
// 解析时间时，默认的月
export var DEFAULT_MONTH = 1;
// 解析时间时，默认的日
export var DEFAULT_DATE = 1;
// 解析时间时，默认的时
export var DEFAULT_HOUR = 0;
// 解析时间时，默认的分
export var DEFAULT_MINUTE = 0;
// 解析时间时，默认的秒
export var DEFAULT_SECOND = 0;
