// @vdian/safety/lib/url-legalize/index.js
import { createLink, htmlDecode, access, isEmpty } from '../util';
import { EMPTRY_URL, WHITE_DOMAINS, ILLEGAL_PROTOCOL_REGEXP } from '../constants';
import { isFunction, merge } from '@vdian/util';
var WHITE_DOMAINS_REGEXP = getDomainsRegexp(WHITE_DOMAINS);
function getRegexpString(domain) {
    // if domain is not start with ., only match like weidian.com
    if (domain[0] !== '.') {
        domain = '^' + domain;
    }
    else {
        domain = '(^|.)' + domain.substr(1);
    }
    // match from end
    domain = domain + '$';
    // replace . to \\.
    domain = domain.replace(/\./g, '\\.');
    return domain;
}
function getDomainsRegexp(domains) {
    var regexp = domains.map(getRegexpString).join('|');
    return new RegExp("(" + regexp + ")");
}
// TODO support kdbridge://weidian.com/ private protocol
// check if protocol contains javascript: or data:
/**
 *
 * @ignore
 */
export function isLegalProtocol(protocol) {
    return !ILLEGAL_PROTOCOL_REGEXP.test(protocol);
}
// check if hostname is in white domains
function isLegalDomain(hostname, options) {
    if (options && options.whiteDomains.length) {
        var domains = getWhiteDomains(options);
        var CUSTOM_WHITE_DOMAINS_REGEXP = getDomainsRegexp(domains);
        return CUSTOM_WHITE_DOMAINS_REGEXP.test(hostname);
    }
    return WHITE_DOMAINS_REGEXP.test(hostname);
}
/**
 *
 * @param options
 * @returns string[] 白名单
 */
function getWhiteDomains(options) {
    var option = merge({ append: true }, options || {});
    var whiteDomains = option.whiteDomains || [];
    // 未设置返回默认
    if (option.append) {
        var domains_1 = WHITE_DOMAINS;
        whiteDomains.forEach(function (domain) {
            domains_1.push(domain);
        });
        return domains_1;
    }
    else {
        return whiteDomains;
    }
}
/**
 * 检查是否是合法的链接，非法或者空链接返回false
 * @param url  链接
 * @param options  设置白名单
 */
export function isLegalUrl(url, options) {
    if (isEmpty(url)) {
        return false;
    }
    // prevent links from appearing in hexadecimal and decimal encoding
    url = htmlDecode(url).trim();
    var link = createLink(url);
    return (isLegalProtocol(link.protocol) && isLegalDomain(link.hostname, options));
}
/**
 * 合法化链接，合法返回绝对路径，不合法返回javascript:;
 * @param url 链接
 * @param options  设置白名单
 * @returns boolean
 */
export function legalizeUrl(url, options) {
    url = isLegalUrl(url, options) ? createLink(url).href : EMPTRY_URL;
    var resolveUrl = access("resolveUrl" /* RESOLVE_URL */);
    if (isFunction(resolveUrl)) {
        url = resolveUrl(url);
    }
    return url;
}
/**
 * 添加域名白名单
 * @param domain 与Cookie Domain设置规则保持一致，例如 .weidian.com 匹配 www.weidian.com,h5.weidian.com，weidian.com 只匹配 weidian.com
 */
export function addWhiteDomain(domain) {
    if (WHITE_DOMAINS.indexOf(domain) === -1) {
        WHITE_DOMAINS.push(domain);
        WHITE_DOMAINS_REGEXP = getDomainsRegexp(WHITE_DOMAINS);
    }
}
