// @vdian/user-agent/commonjs/regex/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// 判断Android系统的正则
exports.ANDROID_REGEXP = /(Android)\s+([\d.]+)/i;
// 判断iOS系统的正则
exports.IOS_REGEXP = /\(i[^;]+;( U;)? cpu.+mac os x/i;
// 判断移动端正则
exports.MOBILE_REGEXP = /AppleWebKit.*Mobile.*/i;
// 判断苹果产品系列正则
exports.APPLE_REGEXP = /(iPhone|iPad|iPod|iOS|Mac OS X)/i;
// 判断iPad正则
exports.IPAD_REGEXP = /(iPad).*OS\s([\d_]+)/i;
// 判断iPhone正则
exports.IPHONE_REGEXP = /(iPhone\sOS)\s([\d_]+)/i;
// 判断QQ APP正则
exports.QQ_REGEXP = /QQ\/([\d.]+)/i;
// 判断微信正则
exports.WEI_XIN_REGEXP = /micromessenger/i;
// 判断企业微信正则
exports.WEI_XIN_WORK_REGEXP = /wxwork\/.* MicroMessenger/i;
// 判断微博正则
exports.WEI_BO_REGEXP = /WeiBo/i;
// 判断买家版正则
exports.WD_BUYER_REGEXP = /WDAPP\(WDBuyer/i;
// 判断卖家版正则
exports.WD_SELLER_REGEXP = /WDAPP\(WD\//i;
// 获取买家版、买家版版本号正则
exports.WD_APP_VERSION_REGEXP = /WDAPP\((?:wdbuyer|wd)\/([\d.]+)/i;
// 判断支付宝正则
exports.ALIPAY_REGEXP = /AlipayClient/i;
// 判断miniProgram正则
exports.MINI_PROGRAM_REGEXP = /miniProgram/i;
