// @vdian/rpx/lib/util/proxy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.proxy = void 0;
var util_1 = require("@vdian/util");
var VENDOR_NAMES_REGEXP = /^(Webkit|Moz|ms)\b/i;
var normalize = util_1.cached(function (prop) {
    prop = util_1.decamelize(prop);
    if (VENDOR_NAMES_REGEXP.test(prop)) {
        prop = '--' + prop;
    }
    return prop;
});
function setAttribute(handler) {
    var originalSetAttribute = Element.prototype.setAttribute;
    Element.prototype.setAttribute = function setAttribute(name, value) {
        if (name === 'style') {
            value = handler(value);
        }
        return originalSetAttribute.call(this, name, value);
    };
}
function getStyleDeclaration() {
    return document.head.style;
}
function defineStyleDeclarationProperty(prototype, name) {
    Object.defineProperty(prototype, name, {
        get: function () {
            return this.getPropertyValue(normalize(name));
        },
        set: function (value) {
            this.setProperty(normalize(name), value);
        }
    });
}
function styleDeclaration(handler) {
    var declaration = getStyleDeclaration();
    var prototype = declaration.__proto__ || window.CSSStyleDeclaration.prototype;
    function setProperty() {
        var originalSetProperty = prototype.setProperty;
        prototype.setProperty = function setProperty(propertyName, value, priority) {
            value = handler(value);
            originalSetProperty.call(this, propertyName, value, priority);
        };
    }
    function declarations() {
        for (var name_1 in declaration) {
            if (declaration.hasOwnProperty(name_1)) {
                defineStyleDeclarationProperty(prototype, name_1);
            }
        }
    }
    if (prototype) {
        setProperty();
        declarations();
    }
}
/**
 * TODO
 * 1. proxy createElement('style')
 * 2. proxy CSSStyleDeclaration.prototype.cssText
 * @ignore
 */
function proxy(handler) {
    try {
        setAttribute(handler);
        styleDeclaration(handler);
    }
    catch (e) { }
}
exports.proxy = proxy;
