// @vdian/safety/commonjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.option = void 0;
var dom_observe_1 = require("./dom-observe");
var util_1 = require("./util");
var escape_1 = require("./escape");
Object.defineProperty(exports, "escape", { enumerable: true, get: function () { return escape_1.escape; } });
var jump_1 = require("./jump");
Object.defineProperty(exports, "jump", { enumerable: true, get: function () { return jump_1.jump; } });
var url_legalize_1 = require("./url-legalize");
Object.defineProperty(exports, "isLegalUrl", { enumerable: true, get: function () { return url_legalize_1.isLegalUrl; } });
Object.defineProperty(exports, "legalizeUrl", { enumerable: true, get: function () { return url_legalize_1.legalizeUrl; } });
Object.defineProperty(exports, "addWhiteDomain", { enumerable: true, get: function () { return url_legalize_1.addWhiteDomain; } });
// 设置默认值
util_1.access({
    observeIframe: true
});
/**
 * 用于配置安全库相关属性
 * @param options 配置属性
 */
function option(options) {
    util_1.access(options);
    // observe or disconnect
    util_1.access("observeDom" /* OBSERVE_DOM */) ? dom_observe_1.observe() : dom_observe_1.disconnect();
}
exports.option = option;
