// @vdian/rpx/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.capture = exports.to = exports.rpxToRem = exports.rpxToPx = void 0;
var util_1 = require("./util");
var util_2 = require("@vdian/util");
var SKETCH_WIDTH = 750;
function getConversionValue() {
    try {
        return Number(process.env.RPX_CONVERSION_VALUE);
    }
    catch (e) { }
}
var CONVERSION_VALUE = getConversionValue() || 100;
function getUnit() {
    try {
        return process.env.RPX_UNIT;
    }
    catch (e) { }
}
var UNIT = getUnit() || 'rem';
function getClientWidth() {
    var width = SKETCH_WIDTH;
    try {
        width = window.document.documentElement.clientWidth;
    }
    catch (e) { }
    return width;
}
/**
 * rpx 转换成 px
 * @deprecated 使用to方法替代，to返回包含转换后的单位，利于集中控制单位。
 * @param value rpx的值
 * @returns px的值
 */
function rpxToPx(value) {
    var width = getClientWidth();
    return (value * width) / SKETCH_WIDTH;
}
exports.rpxToPx = rpxToPx;
/**
 * rpx 转换成 rem
 * @deprecated 使用to方法替代，to返回包含转换后的单位，利于集中控制单位。
 * @param value rpx的值
 * @returns rem 的值
 */
function rpxToRem(value) {
    return value / CONVERSION_VALUE;
}
exports.rpxToRem = rpxToRem;
/**
 * 将rpx转换为通过目标适配单位(目前是rem)计算后的值。
 * @param value rpx的值
 * @returns 转换后的值和单位
 */
function to(value) {
    return value / CONVERSION_VALUE + UNIT;
}
exports.to = to;
/**
 * BE CAREFUL
 * @param options 配置参数
 */
var captured = false;
function capture(options) {
    if (options === void 0) { options = {}; }
    if (captured) {
        return;
    }
    var resolve = options.resolve;
    util_1.proxy(util_2.cached(function (text) {
        return resolve ? resolve(text) : util_1.replaceRpx(text, to);
    }));
    captured = true;
}
exports.capture = capture;
