// @vdian/user-agent/lib/index.js
import { ANDROID_REGEXP, IOS_REGEXP, MOBILE_REGEXP, APPLE_REGEXP, IPAD_REGEXP, IPHONE_REGEXP, QQ_REGEXP, WEI_XIN_REGEXP, WEI_XIN_WORK_REGEXP, WEI_BO_REGEXP, WD_BUYER_REGEXP, WD_SELLER_REGEXP, WD_APP_VERSION_REGEXP, ALIPAY_REGEXP, MINI_PROGRAM_REGEXP } from './regex';
var UA = navigator.userAgent.toLowerCase();
/**
 * 是否是移动端
 */
export function isMobile() {
    return MOBILE_REGEXP.test(UA);
}
/**
 * 是否是Android系统
 */
export function isAndroid() {
    return ANDROID_REGEXP.test(UA);
}
/**
 * 是否是iOS系统
 */
export function isIOS() {
    return IOS_REGEXP.test(UA);
}
/**
 * 是否是苹果产品系列
 */
export function isApple() {
    return APPLE_REGEXP.test(UA);
}
/**
 * 是否是iPhone
 */
export function isIphone() {
    return IPHONE_REGEXP.test(UA);
}
/**
 * 是否是iPad
 */
export function isIpad() {
    return IPAD_REGEXP.test(UA);
}
/**
 * 是否是QQ APP
 */
export function isQQ() {
    return QQ_REGEXP.test(UA);
}
/**
 * 是否是微信
 */
export function isWeiXin() {
    return WEI_XIN_REGEXP.test(UA) && !isWeiXinWork();
}
/**
 * 是否是企业微信
 */
export function isWeiXinWork() {
    return WEI_XIN_WORK_REGEXP.test(UA);
}
/**
 * 是否是微博
 */
export function isWeiBo() {
    return WEI_BO_REGEXP.test(UA);
}
/**
 * 是否是买家版
 */
export function isWDBuyer() {
    return WD_BUYER_REGEXP.test(UA);
}
/**
 * 获取买家版,卖家版APP版本，返回版本号字符串，例如 5.5.1
 * @returns 版本号。 不是买家版、卖家版就返回 undefined
 */
export function getWDAppVersion() {
    var version;
    var match = UA.match(WD_APP_VERSION_REGEXP);
    if (match) {
        version = match[1];
    }
    return version;
}
/**
 * 是否是卖家版
 */
export function isWDSeller() {
    return WD_SELLER_REGEXP.test(UA);
}
/**
 * 判断是否在小程序环境
 */
export function isMiniProgram() {
    return (window.__wxjs_environment === 'miniprogram' ||
        MINI_PROGRAM_REGEXP.test(UA));
}
/**
 * 判断是否在支付宝环境
 */
export function isAlipay() {
    return ALIPAY_REGEXP.test(UA);
}
