// @vdian/safety/commonjs/constants/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FROM_CHAR_CODE = exports.NODE_CACHE_KEY = exports.PARAM_SOURCE_NAMES = exports.ESCAPE_REGEXP = exports.ESCAPE_MAP = exports.EMPTRY_URL = exports.ILLEGAL_PROTOCOL_REGEXP = exports.WHITE_DOMAINS = exports.READY_STATE_REGEXP = exports.A = exports.DOCUMENT = exports.GLOBAL = void 0;
// 全局对象
exports.GLOBAL = window;
// document
exports.DOCUMENT = document;
exports.A = exports.DOCUMENT.createElement('a');
exports.READY_STATE_REGEXP = /complete|loaded|interactive/;
// 白名单域名
exports.WHITE_DOMAINS = [
    '.weidian.com',
    '.koudai.com',
    '.vdian.com',
    '.geilicdn.com',
    '.fangxin.com',
    '.ruyu.com',
    '.vdian.net',
    '.91ruyu.com',
    '.bibikan.cn',
    '.youshop01.com',
    '.youshop02.com',
    '.youshop03.com',
    '.youshop04.com',
    '.youshop05.com',
    '.youshop06.com',
    '.youshop07.com',
    '.youshop08.com',
    '.youshop09.com',
    '.youshop10.com',
    '.kou6ai.cn',
    '.mitao.cn'
];
// 非法的协议头
exports.ILLEGAL_PROTOCOL_REGEXP = /^(javascript|data):/i;
// 空链接
exports.EMPTRY_URL = 'javascript:;';
// escape 字符
exports.ESCAPE_MAP = {
    '&': '&amp;',
    '<': '&lt;',
    '>': '&gt;',
    '"': '&#34;',
    "'": '&#39;',
    '`': '&#96;'
};
// escape 正则，/[&<>"'`]/g
exports.ESCAPE_REGEXP = new RegExp('[' + Object.keys(exports.ESCAPE_MAP).join('') + ']', 'g');
// param标签与资源相关的名称
exports.PARAM_SOURCE_NAMES = ['url', 'src', 'filename'];
// 节点缓存KEY
exports.NODE_CACHE_KEY = '__safetyCache';
exports.FROM_CHAR_CODE = String.fromCharCode;
