// @vdian/user-agent/commonjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var regex_1 = require("./regex");
var UA = navigator.userAgent.toLowerCase();
/**
 * 是否是移动端
 */
function isMobile() {
    return regex_1.MOBILE_REGEXP.test(UA);
}
exports.isMobile = isMobile;
/**
 * 是否是Android系统
 */
function isAndroid() {
    return regex_1.ANDROID_REGEXP.test(UA);
}
exports.isAndroid = isAndroid;
/**
 * 是否是iOS系统
 */
function isIOS() {
    return regex_1.IOS_REGEXP.test(UA);
}
exports.isIOS = isIOS;
/**
 * 是否是苹果产品系列
 */
function isApple() {
    return regex_1.APPLE_REGEXP.test(UA);
}
exports.isApple = isApple;
/**
 * 是否是iPhone
 */
function isIphone() {
    return regex_1.IPHONE_REGEXP.test(UA);
}
exports.isIphone = isIphone;
/**
 * 是否是iPad
 */
function isIpad() {
    return regex_1.IPAD_REGEXP.test(UA);
}
exports.isIpad = isIpad;
/**
 * 是否是QQ APP
 */
function isQQ() {
    return regex_1.QQ_REGEXP.test(UA);
}
exports.isQQ = isQQ;
/**
 * 是否是微信
 */
function isWeiXin() {
    return regex_1.WEI_XIN_REGEXP.test(UA) && !isWeiXinWork();
}
exports.isWeiXin = isWeiXin;
/**
 * 是否是企业微信
 */
function isWeiXinWork() {
    return regex_1.WEI_XIN_WORK_REGEXP.test(UA);
}
exports.isWeiXinWork = isWeiXinWork;
/**
 * 是否是微博
 */
function isWeiBo() {
    return regex_1.WEI_BO_REGEXP.test(UA);
}
exports.isWeiBo = isWeiBo;
/**
 * 是否是买家版
 */
function isWDBuyer() {
    return regex_1.WD_BUYER_REGEXP.test(UA);
}
exports.isWDBuyer = isWDBuyer;
/**
 * 获取买家版,卖家版APP版本，返回版本号字符串，例如 5.5.1
 * @returns 版本号。 不是买家版、卖家版就返回 undefined
 */
function getWDAppVersion() {
    var version;
    var match = UA.match(regex_1.WD_APP_VERSION_REGEXP);
    if (match) {
        version = match[1];
    }
    return version;
}
exports.getWDAppVersion = getWDAppVersion;
/**
 * 是否是卖家版
 */
function isWDSeller() {
    return regex_1.WD_SELLER_REGEXP.test(UA);
}
exports.isWDSeller = isWDSeller;
/**
 * 判断是否在小程序环境
 */
function isMiniProgram() {
    return (window.__wxjs_environment === 'miniprogram' ||
        regex_1.MINI_PROGRAM_REGEXP.test(UA));
}
exports.isMiniProgram = isMiniProgram;
/**
 * 判断是否在支付宝环境
 */
function isAlipay() {
    return regex_1.ALIPAY_REGEXP.test(UA);
}
exports.isAlipay = isAlipay;
