// check-module-latest/index.js
module.exports = ({ moduleName = '', cwd = process.cwd(), registry, } = {}) => {

    if (!moduleName) {
        throw Error('Please pass in moduleName.');
    }

    return new Promise((resolve) => {
        const fs = require('fs');
        const path = require('path');
        const fse = require('fs-extra');

        const getInstalledPath = require('get-installed-path');
        const getNpmPackageVersion = require('get-npm-package-version');

        const childProcess = require('child_process');

        function writeFileSync(filePath, content) {
            fse.ensureFileSync(filePath);
            const fd = fs.openSync(filePath, 'w+');
            fs.writeFileSync(filePath, content);
            fs.close(fd);
        }

        function installModule() {
            if (fs.existsSync(cwd)) {
                fse.removeSync(cwd);
            }

            writeFileSync(path.join(cwd, 'package.json'), JSON.stringify({
                name: 'installing-module',
                version: '1.0.0',
            }));

            try {
                childProcess.execSync(`npm --registry ${registry} install ${moduleName}@latest`, { cwd, stdio: 'inherit' });
            } catch (err) {
                throw Error(err);
            }
        }

        function isLatest(modulePath) {
            let latestVersion = getNpmPackageVersion(moduleName, {
                registry,
                timeout: 2000,
            });

            latestVersion += '';

            const pkgFile = path.join(modulePath, 'package.json');
            let currentVersion = JSON.parse(fs.readFileSync(pkgFile).toString()).version;

            currentVersion += '';

            if (latestVersion && latestVersion !== 'null' && latestVersion !== currentVersion) {
                return false;
            }

            return true;
        }

        // get cmd module successfully
        getInstalledPath(moduleName, { local: true, cwd }).then((modulePath) => {
            resolve({
                isExist: true, 
                isLatest: isLatest(modulePath),
            });
        }).catch(() => {});

        // get cmd module failed
        getInstalledPath(moduleName, { local: true, cwd }).catch(() => {
            resolve({
                isExist: false,
            });
        });
    });
};
