// @vdian/safety/lib/constants/index.js
// 全局对象
export var GLOBAL = window;
// document
export var DOCUMENT = document;
export var A = DOCUMENT.createElement('a');
export var READY_STATE_REGEXP = /complete|loaded|interactive/;
// 白名单域名
export var WHITE_DOMAINS = [
    '.weidian.com',
    '.koudai.com',
    '.vdian.com',
    '.geilicdn.com',
    '.fangxin.com',
    '.ruyu.com',
    '.vdian.net',
    '.91ruyu.com',
    '.bibikan.cn',
    '.youshop01.com',
    '.youshop02.com',
    '.youshop03.com',
    '.youshop04.com',
    '.youshop05.com',
    '.youshop06.com',
    '.youshop07.com',
    '.youshop08.com',
    '.youshop09.com',
    '.youshop10.com',
    '.kou6ai.cn',
    '.mitao.cn'
];
// 非法的协议头
export var ILLEGAL_PROTOCOL_REGEXP = /^(javascript|data):/i;
// 空链接
export var EMPTRY_URL = 'javascript:;';
// escape 字符
export var ESCAPE_MAP = {
    '&': '&amp;',
    '<': '&lt;',
    '>': '&gt;',
    '"': '&#34;',
    "'": '&#39;',
    '`': '&#96;'
};
// escape 正则，/[&<>"'`]/g
export var ESCAPE_REGEXP = new RegExp('[' + Object.keys(ESCAPE_MAP).join('') + ']', 'g');
// param标签与资源相关的名称
export var PARAM_SOURCE_NAMES = ['url', 'src', 'filename'];
// 节点缓存KEY
export var NODE_CACHE_KEY = '__safetyCache';
export var FROM_CHAR_CODE = String.fromCharCode;
