// @vdian/safety/commonjs/util/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isEmpty = exports.toArray = exports.debounce = exports.access = exports.removeChild = exports.htmlDecode = exports.domContentLoaded = exports.createLink = void 0;
var util_1 = require("@vdian/util");
Object.defineProperty(exports, "debounce", { enumerable: true, get: function () { return util_1.debounce; } });
Object.defineProperty(exports, "toArray", { enumerable: true, get: function () { return util_1.toArray; } });
Object.defineProperty(exports, "isEmpty", { enumerable: true, get: function () { return util_1.isEmpty; } });
var constants_1 = require("../constants");
/**
 * @ignore
 */
function createLink(url) {
    constants_1.A.href = url;
    return constants_1.A;
}
exports.createLink = createLink;
/**
 * @ignore
 */
function domContentLoaded(callback) {
    if (constants_1.READY_STATE_REGEXP.test(constants_1.DOCUMENT.readyState) && constants_1.DOCUMENT.body) {
        callback();
        return;
    }
    constants_1.GLOBAL.addEventListener('DOMContentLoaded', callback);
}
exports.domContentLoaded = domContentLoaded;
/**
 * HTML Decode 处理，防止出现10进制或者16进制
 * @param input 字符串
 * @returns 处理后的字符串
 * @ignore
 */
function htmlDecode(input) {
    return input
        .replace(/&#x([\dabcdef]+);?/gi, function (_e, t) {
        return constants_1.FROM_CHAR_CODE(parseInt(t, 16));
    })
        .replace(/&#(\d+);?/g, function (_e, t) {
        return constants_1.FROM_CHAR_CODE(parseInt(t, 10));
    });
}
exports.htmlDecode = htmlDecode;
/**
 * @ignore
 */
function removeChild(node) {
    if (node.parentNode) {
        node.parentNode.removeChild(node);
    }
}
exports.removeChild = removeChild;
var options = {};
/**
 * @ignore
 */
function access(opts) {
    if (util_1.isString(opts)) {
        return options[opts];
    }
    util_1.merge(options, opts);
}
exports.access = access;
